feat: add canister locking example with generic types to security best practices (#2767)
dfinity/portal: docs/developer-docs/security/rust-canister-development-security-best-practices.mdx

- A global lock that does not only lock per caller. For this, set a boolean flag in the canister state instead of using a `BTreeSet<Principal>`.
- A guard that makes sure that only a limited number of principals are allowed to execute a method at the same time. For this, one can return an error in `CallerGuard::new()` in case `pending_requests.len() >= MAX_NUM_CONCURRENT_REQUESTS`.
- A guard that limits the number of times a method can be called in parallel. For this, use a counter in the canister state that is checked and increased in `CallerGuard::new()` and decreased in `Drop`.

> - A lock that uses a different type than `Principal` to grant access to the resource. You can find an implementation using generic types [here](./_attachments/lock.rs).

Reviewer: ```suggestion
- A lock that uses a different type than `Principal` to grant access to the resource. [View an implementation using generic types](./_attachments/lock.rs).
```
